### PR TITLE
Codegen: Generate per-kind reference docs

### PR DESCRIPTION
**Pull request by undef1nd**
**What is this feature?**

Adds a jenny to generate docs for the core structured kind schemas.

Contributes to https://github.com/grafana/schematization-and-as-code-project/issues/28

Fixes https://github.com/grafana/schematization-and-as-code-project/issues/36 as per a2fd0b74b541fbeda54f70e6ce5f8e5e058e0b9a

Fixes https://github.com/grafana/schematization-and-as-code-project/issues/35 as per ceac10b01824d801f147467c428806f03b94bfc0

**Special notes for your reviewer**:
I added two new sections under `Developers` section: `Grafana schema` -> `Core kinds`

### CODE REVIEW

File: docs/sources/developers/kinds/core/team/schema-reference.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+| `avatarUrl`     | string                   | No       | AvatarUrl is the team's avatar URL.                                                                                           |
+| `email`         | string                   | No       | Email of the team.                                                                                                            |
+
+## accessControl
+
+AccessControl metadata associated with a given resource.
+
+| Property | Type | Required | Description |
+|----------|------|----------|-------------|
+
```
Comment (undef1nd) [inline docs/sources/developers/kinds/core/team/schema-reference.md:34]: Does not generate cue's `[string]: bool` properly: same issue as https://github.com/grafana/grafana/pull/59242#discussion_r1031579122

Comment (sdboyer) [reply docs/sources/developers/kinds/core/team/schema-reference.md:34]: Fix for this is now in #60496

**Comment by sdboyer** [comment]
Given that @undef1nd has listed out discovered problems in a set of issues, as long as the generated docs only appear on `next`, then this is good enough to merge and iterate, from my perspective.

Would like @jdbaldry to give his sign off, too

File: .prettierignore [hunk 2]
```diff
@@ -33,3 +33,6 @@ public/openapi3.json
 
 # Generated Kinds report
 kinds/report.json
+
+# Generated schema docs
+docs/sources/developers/kinds/
```
Comment (undef1nd) [inline .prettierignore:38]: I kept getting `prettier` warning reg table formatting. Disabling the prettier for the time being, till I find a better way to render tables in markdown (thus far github.com/olekukonko/tablewriter is being used for that).

This is my diff after running `yarn run prettier:write` on generated docs:
Screenshot 2023-01-10 at 17 15 08

Comment (sdboyer) [reply .prettierignore:38]: Disabling linters on generated artifacts is a common and acceptable practice 👍🏻

Comment (undef1nd) [reply .prettierignore:38]: Thanks for confirming!

File: docs/sources/developers/kinds/_index.md [hunk 3]
```diff
@@ -0,0 +1,9 @@
+---
+draft: true
+title: Grafana schema
+weight: 200
+---
+
+# Grafana schema
+
+[Core kinds]({{< relref "core/_index.md" >}})
```
Comment (rhorvath) [inline docs/sources/developers/kinds/_index.md:9]: Is the Core kinds link working for you? I had to remove the _index.md in my env:
`[Core kinds]({{< relref "core/" >}})`

Comment (joanlopez) [reply docs/sources/developers/kinds/_index.md:9]: It works fine for me!

**Comment by joanlopez** [comment]
Fixed https://github.com/grafana/schematization-and-as-code-project/issues/36 with https://github.com/grafana/grafana/commit/a2fd0b74b541fbeda54f70e6ce5f8e5e058e0b9a (description updated!)

File: kinds/gen.go [hunk 4]
```diff
@@ -40,6 +40,7 @@ func main() {
 		codegen.BaseCoreRegistryJenny(filepath.Join("pkg", "registry", "corekind"), kindsys.GoCoreKindParentPath),
 		codegen.LatestMajorsOrXJenny(kindsys.TSCoreKindParentPath, codegen.TSTypesJenny{}),
 		codegen.TSVeneerIndexJenny(filepath.Join("packages", "grafana-schema", "src")),
+		codegen.DocsJenny(filepath.Join("docs", "sources", "developers", "kinds", "core"), true),
```
Comment (joanlopez) [inline kinds/gen.go:43]: Should we do something similar for composable kinds? (I guess so, but wanna confirm) 🤔  cc/ @sdboyer

I guess would make sense to make `DocsJenny` responsible for both, so we could just pass the path til `.../kinds` and let if create `core` and `composable` directories. Any objection? @undef1nd

**Comment by joanlopez** [comment]
Fixed https://github.com/grafana/schematization-and-as-code-project/issues/35 with ceac10b01824d801f147467c428806f03b94bfc0 (description updated!)

File: kinds/gen.go (hunk 4 above)
Comment (undef1nd) [reply kinds/gen.go:43]: That would be the best, agree.

Comment (sdboyer) [reply kinds/gen.go:43]: yep yep!

In an ideal world, the jenny itself only decides the filename, and there's a generic jenny pipeline component that we use when defining the overall pipeline to specify its leading path.

But the type system for jennies is too clunky right now to really support that, so what you're describing is :+1:. For now, what's really important is having just one jenny, regardless of whether this is a core or composable kind.

File: pkg/codegen/jenny_docs.go [hunk 5]
```diff
@@ -0,0 +1,486 @@
...
+type templateData struct {
+	Draft        bool
+	KindName     string
+	KindVersion  string
+	KindMaturity string
+	Markdown     string
+}
+
+// -------------------- JSON to Markdown conversion --------------------
+// Copied from https://github.com/marcusolsson/json-schema-docs and sliggtly changed to fit the DocsJenny
```
Comment (jdbaldry) [inline pkg/codegen/jenny_docs.go:105]: nit: typo

```suggestion
// Copied from https://github.com/marcusolsson/json-schema-docs and slightly changed to fit the DocsJenny
```

Comment (undef1nd) [reply pkg/codegen/jenny_docs.go:105]: Thanks, Jack!

File: docs/sources/developers/kinds/_index.md [hunk 6]
```diff
@@ -0,0 +1,10 @@
+---
+title: Grafana schema
+weight: 200
+_build:
+  list: false
```
Comment (undef1nd) [inline docs/sources/developers/kinds/_index.md:5]: As discussed with @jdbaldry `draft: true` is not needed anymore so I removed it in this commit https://github.com/grafana/grafana/pull/60416/commits/967cab52fda006bed42e9fd59279a68f22c756a1

**Review by rhorvath — APPROVED** [review]

**Review by joanlopez — APPROVED** [review]
As agreed offline, I'd merge it as-is (looks fine enough for me) and work on the remaining issues later, in individually separated pull requests, easier to digest 🚀

File: pkg/codegen/jenny_docs.go [hunk 7]
```diff
@@ -0,0 +1,490 @@
...
+			Schemas json.RawMessage
+		}
+	}
+	err = json.Unmarshal(b, &obj)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal schema json: %v", err)
+	}
+
+	// fixes the references between the types within a json after making components.schema.<types> the root of the json
+	kindJsonStr := strings.Replace(string(obj.Components.Schemas), "#/components/schemas/", "#/", -1)
```
[the remark this answers is not in the document]
Comment (undef1nd) [reply pkg/codegen/jenny_docs.go:59]: Yep, right!

**Pull request merged**